benchmark/scenarios/write_request_load.py

```diff
@@ -0,0 +1,288 @@
+# Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
+"""
+Read request load scenario for the control service benchmarks.
+"""
+from collections import deque
+from itertools import repeat
+
+from zope.interface import implementer
+import eliot
+
+from twisted.internet.defer import CancelledError, Deferred, succeed
+from twisted.internet.task import LoopingCall
+
+from flocker.common import loop_until, timeout
+
+from .._interfaces import IScenario
+
+DEFAULT_SAMPLE_SIZE = 5
+
+# XXX make it a common class for Read and Write scenarios.
+# Note that the docstrings are not up-to-date because they are being updated
+# in the Read scenario
+class WRateMeasurer(object):
+    """
+    Measures the rate of requests in requests per second
+
+    :ivar sample_size: size of the sample we request - how many samples,
+    or counts, do we want to consider we have reached the rate.
+    """
+
+    def __init__(self, sample_size=DEFAULT_SAMPLE_SIZE):
+        self.counts = deque([0] * sample_size, sample_size)
+        self.sent = 0
+        self.received = 0
+        self._rate = 0
+        self.sample_size = sample_size
+
+    def send_request(self):
+        """
+        Increase the counter of sent requests.
+        """
+        self.sent += 1
+
+    def receive_request(self, result):
+        """
+        Increase the counter of sent requests.
+        """
+        self.received += 1
+
+    def update_rate(self):
+        """
+        Update the current rate and stores a new count in the counts list.
+        """
+        self._rate = (self.received - self.counts[0]) / float(self.sample_size)
+        self.counts.append(self.received)
+
+    def outstanding(self):
+        """
+        Return the number of outstanding requests.
+        """
+        return self.sent - self.received
+
+    def rate(self):
+        return self._rate
+
+
+# XXX make this excetions common for read and write scenarios and remove
+# the initial W preceding the name
+class WRequestRateTooLow(Exception):
+    """
+    The RequestRate dropped below a threshold.
+    """
+
+
+class WRequestRateNotReached(Exception):
+    """
+    The RequestRate did not reach the target level.
+    """
+
+
+class WRequestOverload(Exception):
+    """
+    There are too many outstanding request.
+    """
+
+class WDataseCreationTimeout(Exception):
+    """
+    The dataset could not be created.
```

Comment:
It's probably worth mentioning the timeout here.
Although the class name is already quite self-descriptive.